File: Homework01/rsa.go
package main

import "math"

import "math/rand"
import "math/big"
import "errors"
import "fmt"

type Key struct {
    key int
    n int
}

type KeyPair struct {
    Private Key
    Public Key
}

func isPrime(n int) bool {
    if n == 1 {
        return false
    }


    for i := 2; i <= int(math.Floor(math.Sqrt(float64(n)))); i++ {
        if n % i == 0 {
            return false
        }
    }
    return true
}


func gcd(a int, b int) int {
    for {
      if(a>b){
        a%=b;
      }else{
        b%=a;}
      if (a*b==0){return (a|b);}
    }
  }



func multiplicativeInverse(e int, phi int) int {
    var table [][]int
    A := phi
    B := e
    row := []int{A, B, A % B, A / B, -1, -1}
    table = append(table, row)
    for i := 0; table[i][2] != 0; i++ {
        A = table[i][1]
        B = table[i][2]
        row := []int{A, B, A % B, A / B, -1, -1}
        table = append(table, row)
    }

    table[len(table)-1][4] = 0
    table[len(table)-1][5] = 1
    for i := len(table)-2; i >= 0; i-- {
        table[i][4] = table[i+1][5]
        table[i][5] = table[i+1][4] - table[i+1][5]*table[i][3]
    }

    r := table[0][5] % phi
    if r < 0 {
        r = r + phi
    }
    return r
}


func GenerateKeypair(p int, q int) (KeyPair, error) {
    if !(isPrime(p) && isPrime(q)){
        return KeyPair{}, errors.New("Both numbers must be prime.")
    } else if  p == q {
        return KeyPair{}, errors.New("p and q can't be equal.");
    }

    n := p * q
    phi := (p - 1)*(q - 1)
    e := rand.Intn(phi - 1) + 1
    g := gcd(e, phi)
    for g != 1 {
        e = rand.Intn(phi - 1) + 1
        g = gcd(e, phi)
    }

    d := multiplicativeInverse(e, phi)
    return KeyPair{Key{e, n}, Key{d, n}}, nil
}


func Encrypt(pk Key, plaintext string) []int {
    cipher := []int{}
    n := new(big.Int)
    for _, ch := range plaintext {
        n = new(big.Int).Exp(
            big.NewInt(int64(ch)), big.NewInt(int64(pk.key)), nil)
        n = new(big.Int).Mod(n, big.NewInt(int64(pk.n)))
        cipher = append(cipher, int(n.Int64()))
    }
    return cipher
}


func Decrypt(pk Key, cipher []int) string {
    plaintext := ""
    n := new(big.Int)
    for _, ch := range cipher {
        n = new(big.Int).Exp(
            big.NewInt(int64(ch)), big.NewInt(int64(pk.key)), nil)
        n = new(big.Int).Mod(n, big.NewInt(int64(pk.n)))
        plaintext += string(rune(int(n.Int64())))
    }
    return plaintext
}
func main(){
  var q, p int;

  fmt.Print("Enter a number: ")
  fmt.Scan(&q)
  fmt.Print("Enter another number: ")
  fmt.Scan(&p)
  //var pk KeyPair
  pk, _:= GenerateKeypair(p,q)
  var text string
  fmt.Scan(&text)
  var encrypted []int
  encrypted = Encrypt(pk.Private,text)
  var decrypted string
  decrypted = Decrypt(pk.Public, encrypted)
  fmt.Println(encrypted)
  fmt.Println(decrypted)

}

File: Homework01/rsa.py
from math import sqrt
import random


def is_prime(n: int) -> bool:
    """
    >>> is_prime(2)
    True
    >>> is_prime(11)
    True
    >>> is_prime(8)
    False
    """
    if n == 1:
        return False
    for i in range(2, int(sqrt(n)) + 1):
        if not n % i:
            return False
    return True


def gcd(a: int, b: int) -> int:
    """
    >>> gcd(12, 15)
    3
    >>> gcd(3, 7)
    1
    """
    while a and b:
        if a > b:
            a %= b
        else:
            b %= a
    return a | b


def extended_gcd(a: int, b: int):
    if a == 0:
        return b, 0, 1
    else:
        g, y, x = extended_gcd(b % a, a)
        return g, x - (b // a) * y, y


def multiplicative_inverse(e: int, phi: int) -> int:
    """
    >>> multiplicative_inverse(7, 40)
    23
    """
    g, x, y = extended_gcd(e, phi)
    return x % phi


def generate_keypair(p: int, q: int):
    if not (is_prime(p) and is_prime(q)):
        raise ValueError('Both numbers must be prime.')
    elif p == q:
        raise ValueError('p and q cannot be equal')

    n = p * q

    phi = (p-1) * (q-1)
    e = random.randrange(1, phi)
    g = gcd(e, phi)
    while g != 1:
        e = random.randrange(1, phi)
        g = gcd(e, phi)
    d = multiplicative_inverse(e, phi)
    return ((e, n), (d, n))


def encrypt(pk, plaintext):
    key, n = pk
    cipher = [(ord(char) ** key) % n for char in plaintext]
    return cipher


def decrypt(pk, ciphertext):
    key, n = pk
    plain = [chr((char ** key) % n) for char in ciphertext]
    return ''.join(plain)


if __name__ == "__main__":
    p = int(input("Enter a prime number: "))
    q = int(input("Enter ANOTHER prime number: "))
    public, private = generate_keypair(p, q)
    print ("Your public key is ", public, " and your private key is ", private)
    message = input("Enter a message to encrypt with your private key: ")
    encrypted_msg = encrypt(private, message)
    print (encrypted_msg)
    print (decrypt(public, encrypted_msg))
